packages/test-base/src/androidAndroidTest/kotlin/io/realm/kotlin/test/shared/SerializationTests.kt

```diff
@@ -0,0 +1,416 @@
+import io.realm.kotlin.serializers.kotlinxserializers.MutableRealmIntKSerializer
+import io.realm.kotlin.serializers.kotlinxserializers.RealmAnyKSerializer
+import io.realm.kotlin.serializers.kotlinxserializers.RealmInstantKSerializer
+import io.realm.kotlin.serializers.kotlinxserializers.RealmListKSerializer
+import io.realm.kotlin.serializers.kotlinxserializers.RealmSetKSerializer
+import io.realm.kotlin.serializers.kotlinxserializers.RealmUUIDKSerializer
+import io.realm.kotlin.test.GenericTypeSafetyManager
+import io.realm.kotlin.test.platform.PlatformUtils
+import io.realm.kotlin.test.util.TypeDescriptor
+import io.realm.kotlin.types.EmbeddedRealmObject
+import io.realm.kotlin.types.ObjectId
+import io.realm.kotlin.types.RealmAny
+import io.realm.kotlin.types.RealmDictionary
+import io.realm.kotlin.types.RealmInstant
+import io.realm.kotlin.types.RealmObject
+import io.realm.kotlin.types.RealmUUID
+import kotlinx.serialization.UseSerializers
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.polymorphic
+import kotlinx.serialization.modules.subclass
+import org.mongodb.kbson.BsonObjectId
+import org.mongodb.kbson.Decimal128
+import kotlin.reflect.KClass
+import kotlin.reflect.KClassifier
+import kotlin.reflect.KMutableProperty1
+import kotlin.test.AfterTest
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertTrue
+import kotlin.test.fail
+
+class SerializationTests {
+    private lateinit var tmpDir: String
+    private lateinit var realm: Realm
+
+    private lateinit var configuration: RealmConfiguration
+
+    private val json = Json {
+        serializersModule = SerializersModule {
+            polymorphic(RealmObject::class) {
+                subclass(SerializableSample::class)
+            }
+
+            polymorphic(EmbeddedRealmObject::class) {
+                subclass(SerializableEmbeddedObject::class)
+            }
+        }
+    }
+
+    @BeforeTest
+    fun setup() {
+        tmpDir = PlatformUtils.createTempDir()
+        configuration = RealmConfiguration
+            .Builder(setOf(SerializableSample::class, SerializableEmbeddedObject::class))
+            .directory(tmpDir)
+            .build()
+        realm = Realm.open(configuration)
+    }
+
+    @AfterTest
+    fun tearDown() {
+        if (this::realm.isInitialized && !realm.isClosed()) {
+            realm.close()
+        }
+        PlatformUtils.deleteTempDir(tmpDir)
+    }
+
+    private val OBJECT_VALUES = listOf(
+        SerializableSample().apply { stringField = "A" },
+        SerializableSample().apply { stringField = "B" }
+    )
+
+    private fun Collection<TypeDescriptor.RealmFieldType>.mapCollectionDataSets(
```

Comment:
It is extremely difficult to follow the overall exhaustive pattern of these tests. I don't know how to improve when we have to iterate both the collection types, classifiers and nullability though. I definitely don't think we should rework it now, but I think we should be open for alternative ways of doing this, e.g. generating smaller tests that are easy to reason about, etc. 

But in it's current form we should at least add comments about what these utility methods do, to save us all from de-compile the concepts from the code.

Reply:
Yes, the exhaustive test platform is a little bit convoluted. I am going to add some comments as suggested.